### DISCUSSION TITLE
get cell connection data from grb for disv grid.

### DISCUSSION THREAD

**Discussion by RyanConway91** [modflowpy/flopy#1682] [Q&A] [answered]
I want to know my grid cell connections for a disv grid. When I load the grb file (MfGrdFile(grb_file)) I am not seeing any properties that give me the grid cell connections. Is this information in there somewhere in the grb class object? If so, how do I access?

I am trying to recreate the 'targpestu' interpolator and need the cell connections to do the distance weighting.

**Comment by jlarsen-usgs** [comment, marked as answer]
@RyanConway91 

The cell connection information can be accessed and assembled using the `ia` and `ja` property methods on the MfGridFile object. Alternatively, you can get a modelgrid object from MfGrdFile and then use the `.neighbors()` method to get grid cell connections.

Example:
```python
grb = MfGridFile("my_model.dis.grb")
modelgrid = grb.modelgrid

cell_connections = modelgrid.neighbors(0, 1, 2)  # cellid can be input as lay, row, col; lay, node; or node 
```

**Reply by RyanConway91** [reply]
Thank you. 
modelgrid.neighbors() does not run on my model (6 layers, 100000 cells per layer). It just hangs forever with no output. How can I parse the connection data from ia and ja? Sorry, I tried looking at the documentation (https://flopy.readthedocs.io/en/3.3.6/source/flopy.mf6.utils.binarygrid_util.html) but the property descriptions are not helpful.

Here is a link to my grb if you are interested
https://drive.google.com/file/d/1FGUUXwMoiPIJM0tsQIlmTGP87aV1cIZ9/view?usp=share_link

**Reply by jlarsen-usgs** [reply]
@RyanConway91 

The reason it is hanging is that the modelgrid object is assembling all of the connections for your model. It only needs to do this once, but it sounds like this takes awhile for very large grids.

**For the other approach: Note, edited based on solution below** 
I agree that the documentation needs to be updated. The descriptions for those parameters in the `MfGrdFile` class could use some improvement.

`.ia` is an array that defines the indexes that represent the number of connections + 1 per cell. This is a flat array and can be used to index the jagged connection array.
`.ja` is a representation of the jagged connection array in your model. The ja array lists the cell number and the cells connected to it.
As code:

```python
grb = MfGrdFile("mv.dis.grb")
ia = grb.ia
ja = grb.ja
connections = {}
for cell in range(grb.ncells):
    i0 = ia[cell] + 1
    i1 = ia[cell + 1]
    ncon = i1 - i0
    conn = ja[i0:i1]
    connections[cell] = (ncon, conn)
```

**Reply by christianlangevin** [reply]
The IA and JA arrays, including approaches for using them to loop through connections, are described in `mf6io.pdf` in the section called "Description of Binary Output Files for the Groundwater Flow (GWF) and Groundwater Transport (GWT) Models".

**Reply by RyanConway91** [reply]
For some reason that code threw an error for me. Looking at the document (thanks @langevin-usgs) I came up with this and it seems to work based on some visual review:

```
ja = grb.ja
ia = grb.ia
connections = {}#{node:(nconn,[conn_1,...,conn_n])}
ncells = grb.ncells
for n in range(ncells):
    ncon = ia[n+1]-ia[n]-1
    cons = []
    if ncon < 0:
        ncon = 0
    for ipos in range(ia[n]+1,ia[n+1]):
        m = ja[ipos]
        cons.append(m)
    connections[n] = (ncon,cons)
```
